pkg/controller/master/node/promote_job.go

```diff
@@ -0,0 +1,161 @@
+package node
+
+import (
+	"fmt"
+
+	"github.com/rancher/wrangler/pkg/condition"
+	"github.com/rancher/wrangler/pkg/name"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	promoteJobNamespaceName = "harvester-system"
+
+	promoteImage           = "busybox:1.32.0"
+	promoteImagePullPolicy = corev1.PullIfNotPresent
+	promoteRootMountPath   = "/host"
+
+	// restart after modify the k3s service profile
+	promoteCommand = `echo start promote && \
+if [ ! -f /var/lib/rancher/k3os/config.yaml ]; then \
+	sudo cp /k3os/system/config.yaml /var/lib/rancher/k3os/config.yaml && \
+	echo clone config; \
+fi && \
+echo update config && \
+sudo sed -i 's/agent/server/g' /var/lib/rancher/k3os/config.yaml && \
+sudo sed -i 's/command_args=\"agent/command_args=\"server/g' /etc/init.d/k3s-service && \
+echo restart k3s && \
+cat /var/run/k3s-restarter-trap.pid | xargs -r kill -HUP && \
+echo finish promote
+`
+)
+
+var (
+	promotePrivileged   = true
+	promoteBackoffLimit = int32(2)
+
+	ConditionJobComplete = condition.Cond(batchv1.JobComplete)
+	ConditionJobFailed   = condition.Cond(batchv1.JobFailed)
+)
+
+func buildPromoteJobName(nodeName string) string {
+	return name.SafeConcatName("harvester", "promote", nodeName)
+}
+
+func (h *Handler) createPromoteJob(node *corev1.Node) (*batchv1.Job, error) {
+	job := buildPromoteJob(node.Name, promoteImage, []string{"sh"}, []string{"-c",
+		fmt.Sprintf(`chroot %s bash -c "%s"`, promoteRootMountPath, promoteCommand)})
+	j, err := h.jobs.Create(job)
+	return j, checkError(fmt.Sprintf("create job %s on node %s", job.Name, node.Name), err)
+}
+
+func (h *Handler) deleteJob(job *batchv1.Job, deletionPropagation metav1.DeletionPropagation) error {
+	err := h.jobs.Delete(job.Namespace, job.Name, &metav1.DeleteOptions{PropagationPolicy: &deletionPropagation})
+	return checkError(fmt.Sprintf("delete job %s", job.Name), err)
+}
+
+func buildPromoteJob(nodeName string, image string, command, args []string) *batchv1.Job {
```

Comment:
If image, command, and args are constants, we don't need to pass them as parameters